elixir/apps/web/lib/web/controllers/oidc_controller.ex

```
@@ -340,59 +340,129 @@ defmodule Web.OIDCController do
-        SELECT id FROM actors
-        WHERE account_id = $1 AND email = $2 AND disabled_at IS NULL AND $3 = true
-        LIMIT 1
-      ),
-      existing_identity AS (
-        SELECT ei.actor_id
-        FROM external_identities ei
-        JOIN actors a ON a.id = ei.actor_id
-        WHERE ei.account_id = $1 AND ei.issuer = $4 AND ei.idp_id = $5
-          AND a.disabled_at IS NULL
-        LIMIT 1
-      )
-      INSERT INTO external_identities (
-        id, account_id, issuer, idp_id, actor_id,
-        #{insert_fields},
-        inserted_at
-      )
-      SELECT uuid_generate_v4(), $1, $4, $5,
-             COALESCE(actor_lookup.id, existing_identity.actor_id),
-             #{value_placeholders}, $15
-      FROM (SELECT 1) AS dummy
-      LEFT JOIN actor_lookup ON true
-      LEFT JOIN existing_identity ON true
-      WHERE actor_lookup.id IS NOT NULL OR existing_identity.actor_id IS NOT NULL
-      ON CONFLICT (account_id, idp_id, issuer)
-      DO UPDATE SET #{update_set}
-      RETURNING *
-      """
-
-      params =
-        [Ecto.UUID.dump!(account_id), email, email_verified, issuer, idp_id] ++
-          profile_values ++ [DateTime.utc_now()]
-
-      case Safe.unscoped() |> Safe.query(query, params) do
-        {:ok, %{rows: [row], columns: cols}} ->
-          identity = Safe.load(ExternalIdentity, {cols, row})
+      now = DateTime.utc_now()
+
+      profile_fields = [
+        :email,
+        :name,
+        :given_name,
+        :family_name,
+        :middle_name,
+        :nickname,
+        :preferred_username,
+        :profile,
+        :picture
+      ]
+
+      # Convert account_id to the DB (binary) format expected by uuid/binary_id columns
+      account_id_db =
+        case account_id do
+          # already a 16-byte binary
+          <<_::128>> -> account_id
+          # string UUID -> dump to binary
+          _ -> Ecto.UUID.dump!(account_id)
+        end
+
+      # Ensure profile_attrs is at least an empty map
```

I think this is unnecessary? Will always be sent by the caller.

---

are you saying there's no chance that the arg passed in will be nil?

---

If so, perhaps we should update the function definition to be:
```
def upsert_identity(account_id, email, email_verified, issuer, idp_id, %{} = profile_attrs)
```

---

That comes directly from the `claims` and `userinfo` a couple private function calls up the callstack - those come from the `Web.OIDC.verify_token` and `Web.OIDC.fetch_userinfo` which only return maps (or fail) so we should be ok.

---

Ok, then I'll update the function with the last arg match so that if there is any reason we don't get a map then it will fail loudly.